evap/staff/tools.py

```diff
@@ -68,31 +69,95 @@ def delete_navbar_cache_for_users(users):
+
+
+def bulk_update_users(request, user_file, test_run):
+    # pylint: disable=too-many-branches
+    # user_file must have one user per line in the format "{username},{email}"
+    imported_emails = {clean_email(line.decode().split(',')[1]) for line in user_file.readlines()}
+
+    emails_of_users_to_be_created = []
+    users_to_be_updated = []
+    emails_of_non_obsolete_users = list(imported_emails)
+    skipped_external_emails_counter = 0
+
+    for imported_email in imported_emails:
+        if is_external_email(imported_email):
+            skipped_external_emails_counter += 1
+            continue
+        matching_user = find_matching_internal_user_for_email(request, imported_email)
+        if not matching_user:
+            emails_of_users_to_be_created.append(imported_email)
+        elif matching_user.email != imported_email:
+            users_to_be_updated.append((matching_user, imported_email))
+            emails_of_non_obsolete_users.append(matching_user.email)  # add old address so this user won't be deleted
+
+    obsolete_users = UserProfile.objects.exclude(email__in=emails_of_non_obsolete_users)
+    deletable_users = [user for user in obsolete_users if user.can_be_deleted_by_manager]
+    users_to_mark_inactive = [
+        user for user in obsolete_users
+        if user not in deletable_users and user.is_active and user.can_be_marked_inactive_by_manager
+    ]
+
+    messages.info(
+        request,
+        _('The uploaded text file contains {} internal and {} external users. The external users will be ignored. '
+        '{} users are currently in the database. Of those, {} will be updated, {} will be deleted and {} will be '
+        'marked inactive. {} new users will be created.')
+        .format(len(imported_emails)-skipped_external_emails_counter, skipped_external_emails_counter,
+            UserProfile.objects.count(), len(users_to_be_updated), len(deletable_users), len(users_to_mark_inactive),
+            len(emails_of_users_to_be_created))
+    )
+    if users_to_be_updated:
+        messages.info(
+            request,
+            mark_safe(
+                _('Users to be updated are:<br />{}')
+                .format('<br />'.join(
+                    [_("{} > {}").format(user.email, email) for user, email in users_to_be_updated]
+                ))
+            )
+        )
+    if deletable_users:
+        messages.info(request, mark_safe(_('Users to be deleted are:<br />{}')
+            .format('<br />'.join([user.email or user.username for user in deletable_users]))))
+    if users_to_mark_inactive:
+        messages.info(request, mark_safe(_('Users to be marked inactive are:<br />{}')
+            .format('<br />'.join([user.email or user.username for user in users_to_mark_inactive]))))
+    if emails_of_users_to_be_created:
+        messages.info(request, mark_safe(_('Users to be created are:<br />{}')
+            .format('<br />'.join(emails_of_users_to_be_created))))
```

Comment:
Those `mark_safe` should be replaced by [`format_html_join`](https://docs.djangoproject.com/en/2.2/ref/utils/#django.utils.html.format_html_join). If it works, haven’t tested it yet :D